File: ts/src/analyzer/reference-call-receiver.test.ts
import { describe, expect, test } from "vitest";

import type { AstNode } from "../ir/primitive/ast-node.js";
import { AST_TYPE } from "../parser/ast-type.js";
import { referenceCallReceiverFlags } from "./reference-call-receiver.js";

const node = (type: string): AstNode => ({ type }) as unknown as AstNode;

describe("referenceCallReceiverFlags", () => {
  test("CallExpression#callee → call only", () => {
    expect(
      referenceCallReceiverFlags(node(AST_TYPE.CallExpression), "callee"),
    ).toEqual({ call: true, receiver: false });
  });

  test("NewExpression#callee → call only", () => {
    expect(
      referenceCallReceiverFlags(node(AST_TYPE.NewExpression), "callee"),
    ).toEqual({ call: true, receiver: false });
  });

  test("MemberExpression#object → receiver only", () => {
    expect(
      referenceCallReceiverFlags(node(AST_TYPE.MemberExpression), "object"),
    ).toEqual({ call: false, receiver: true });
  });

  test("CallExpression at non-callee key → neither", () => {
    expect(
      referenceCallReceiverFlags(node(AST_TYPE.CallExpression), "arguments"),
    ).toEqual({ call: false, receiver: false });
  });

  test("MemberExpression at non-object key → neither", () => {
    expect(
      referenceCallReceiverFlags(node(AST_TYPE.MemberExpression), "property"),
    ).toEqual({ call: false, receiver: false });
  });

  test("null parent → neither", () => {
    expect(referenceCallReceiverFlags(null, null)).toEqual({
      call: false,
      receiver: false,
    });
  });

  test("unrelated parent type → neither", () => {
    expect(
      referenceCallReceiverFlags(node(AST_TYPE.VariableDeclarator), "init"),
    ).toEqual({ call: false, receiver: false });
  });
});

File: crates/unsnarl-analyzer/src/reference_call_receiver_test.rs
use unsnarl_oxc_parity::AstType;

use super::reference_call_receiver_flags;

#[test]
fn call_expression_callee_marks_call_only() {
    let f = reference_call_receiver_flags(Some(&AstType::CallExpression), Some("callee"));
    assert!(f.call);
    assert!(!f.receiver);
}

#[test]
fn new_expression_callee_marks_call_only() {
    let f = reference_call_receiver_flags(Some(&AstType::NewExpression), Some("callee"));
    assert!(f.call);
    assert!(!f.receiver);
}

#[test]
fn member_expression_object_marks_receiver_only() {
    let f = reference_call_receiver_flags(Some(&AstType::MemberExpression), Some("object"));
    assert!(!f.call);
    assert!(f.receiver);
}

#[test]
fn call_expression_non_callee_is_neither() {
    let f = reference_call_receiver_flags(Some(&AstType::CallExpression), Some("arguments"));
    assert!(!f.call);
    assert!(!f.receiver);
}

#[test]
fn member_expression_non_object_is_neither() {
    let f = reference_call_receiver_flags(Some(&AstType::MemberExpression), Some("property"));
    assert!(!f.call);
    assert!(!f.receiver);
}

#[test]
fn null_parent_is_neither() {
    let f = reference_call_receiver_flags(None, None);
    assert!(!f.call);
    assert!(!f.receiver);
}

#[test]
fn unrelated_parent_type_is_neither() {
    let f = reference_call_receiver_flags(Some(&AstType::VariableDeclarator), Some("init"));
    assert!(!f.call);
    assert!(!f.receiver);
}
